File: go/special_int.go
package dgo

type _SpecialIntKind uint64

const (
	sikDartCallback _SpecialIntKind = iota
	sikGoCallback
	sikGoObject
	sikGoMethod
	sikDartFutureCallback
	sikDartStreamCallback
	sikDartCallbackGroup
	sikPreservedGoCall
)

const (
	specialIntMagic uint64 = 0b0111_1111_1111_1
)

type _SpecialInt interface {
	specialInt()
}

type _Serializable interface {
	_SpecialInt
	getPayload() uint64
	getKind() _SpecialIntKind
}

type _Handlable interface {
	_SpecialInt
	handleCObjects(objs []*Dart_CObject)
}

func _buildSpecialInt(kind _SpecialIntKind, port *Port, payload uint64) _SpecialInt {
	switch kind {
	case sikDartCallback:
		return DartCallback{payload, port}
	case sikDartFutureCallback:
		return DartFutureCallback{payload, port}
	case sikDartStreamCallback:
		return DartStreamCallback{payload, port}
	default:
		panic("unreachable")
	}
}

func _buildHandlable(kind _SpecialIntKind, port *Port, payload uint64) _Handlable {
	switch kind {
	case sikGoCallback:
		return invokingGoCallback{payload, port}
	case sikGoMethod:
		return invokingGoMethod{payload, port}
	case sikPreservedGoCall:
		return _PreservedGoCall{payload, port}
	default:
		panic("unreachable")
	}
}

func parseSpecialInt(port *Port, value uint64, handlable bool) (result _SpecialInt, success bool) {
	if value>>(64-13) != specialIntMagic {
		return nil, false
	}
	kind := _SpecialIntKind((value >> 48) & 0b111)
	payload := value & ((1 << 48) - 1)

	if handlable {
		result = _buildHandlable(kind, port, payload)
	} else {
		result = _buildSpecialInt(kind, port, payload)
	}
	return result, true
}

func serialize(kind _SpecialIntKind, payload uint64) uint64 {
	return (specialIntMagic << (64 - 13)) |
		(uint64(kind) << 48) |
		payload
}

func serializeSpecialInt(i _Serializable) uint64 {
	return serialize(i.getKind(), i.getPayload())
}

File: dart/lib/src/special_int.dart
part of dgo;

enum _SpecialIntKind {
  dartCalback(0),
  goCallback(1),
  goObject(2),
  goMethod(3),
  dartFutureCallback(4),
  dartStreamCallback(5),
  dartCallbackGroup(6),
  prevseredGoCall(7);

  const _SpecialIntKind(this.value) : assert(value >= 0 && value <= 7);
  factory _SpecialIntKind.fromInt(int value) {
    return _SpecialIntKind.values[value];
  }
  final int value;
}

abstract class _SpecialInt {}

// magic   (13 bits) = 0111_1111_1111_1
// kind    (3 bits)
// payload (48 bits)
abstract class _Serializable extends _SpecialInt {
  int get _payload;
  _SpecialIntKind get _kind;

  static final _payloadMask = 0xFFFFFFFFFFFF; // 48 bits
  static const _magic = 0x0FFF; // 13 bits
}

abstract class _Handlable extends _SpecialInt {
  void _handleObjects(Iterable objs);
}

extension _SerializableExt on _Serializable {
  num serialize({required bool asDouble}) {
    final payload = _payload;
    assert(
      payload & _Serializable._payloadMask == payload,
      'dgo:dart: specialInt overflow (max 48 bits)',
    );

    final value =
        (_Serializable._magic << (64 - 13)) | (_kind.value << 48) | payload;

    return asDouble ? value.bitsAsFloat64 : value;
  }
}

typedef _Builder<R extends _SpecialInt> = R Function(int payload, DgoPort port);

_Builder<_SpecialInt> _kindAsBuilder(_SpecialIntKind kind) {
  switch (kind) {
    case _SpecialIntKind.dartCalback:
      return DartCallback._;
    case _SpecialIntKind.goCallback:
      return GoCallback._;
    default:
      throw 'unreachable';
  }
}

_Builder<_Handlable> _kindAsHandlableBuilder(_SpecialIntKind kind) {
  switch (kind) {
    case _SpecialIntKind.dartCalback:
      return _InvokingDartCallback._;
    case _SpecialIntKind.dartCallbackGroup:
      return _DartCallbackGroup._;
    default:
      throw 'unreachable';
  }
}

extension _NumParseExt on num {
  R? _parse<R extends _SpecialInt>(
    DgoPort port,
    _Builder<R> Function(_SpecialIntKind) getBuilder,
  ) {
    final intValue = bitsAsUint64;

    if ((intValue >>> (64 - 13)) != _Serializable._magic) return null;

    final kind = _SpecialIntKind.fromInt((intValue >>> 48) & 0x7);
    final payload = intValue & ((1 << 48) - 1);

    return getBuilder(kind)(payload, port);
  }

  _SpecialInt? parse(DgoPort port) => _parse(port, _kindAsBuilder);

  _Handlable? parseHandlable(DgoPort port) =>
      _parse(port, _kindAsHandlableBuilder);
}
